From a7f66d0b6d577e6f2bb3bf130505643cc761dc81 Mon Sep 17 00:00:00 2001
From: Mohan Kumar G <quic_mkumarg@quicinc.com>
Date: Fri, 22 Dec 2023 12:17:10 +0530
[CAF Only] Subject: [PATCH] hostapd: Fix VHT center frequency for EHT 240 MHz

Currently, when 240 MHz is configured in 5GHz band, the VHT center
frequency is not set correctly.

This is because in hostapd_set_oper_centr_freq_seg0_idx, the
center frequency is not calculated for 5GHz 240MHz case. So, it's
frequency is set the same as 320MHz.

When radar is detected, the punct_update_legacy_bw receives bandwidth
of 160MHz but center frequency for 320MHz. This causes wrong
calculation of new channel bandwidth during puncturing.

Fix this issue by setting the correct center freqency and getting
it from hostapd_get_oper_centr_freq_seg0_idx instead of directly
giving VHT frequency.

Signed-off-by: Mohan Kumar G <quic_mkumarg@quicinc.com>
---
 src/ap/ap_config.h      |  2 ++
 src/ap/dfs.c            |  4 ++--
 src/ap/ieee802_11_vht.c | 11 ++---------
 3 files changed, 6 insertions(+), 11 deletions(-)

--- a/src/ap/ap_config.h
+++ b/src/ap/ap_config.h
@@ -1285,6 +1285,14 @@ hostapd_set_oper_centr_freq_seg0_idx(str
 	if (center_idx_to_bw_6ghz(oper_centr_freq_seg0_idx) == 4)
 		oper_centr_freq_seg0_idx +=
 			conf->channel > oper_centr_freq_seg0_idx ? 16 : -16;
+
+	if (conf->eht_oper_chwidth == CONF_OPER_CHWIDTH_320MHZ &&
+	    oper_centr_freq_seg0_idx == 130) {
+		if (conf->channel < oper_centr_freq_seg0_idx)
+			oper_centr_freq_seg0_idx = oper_centr_freq_seg0_idx - 16;
+		else
+			oper_centr_freq_seg0_idx = oper_centr_freq_seg0_idx + 8;
+	}
 #endif /* CONFIG_IEEE80211BE */
 #ifdef CONFIG_IEEE80211AX
 	if (conf->ieee80211ax)
--- a/src/ap/dfs.c
+++ b/src/ap/dfs.c
@@ -1763,8 +1763,8 @@ int hostapd_dfs_radar_detected(struct ho
 	if (iface->conf->use_ru_puncture_dfs && hostapd_is_usable_ru_punct_bitmap(iface)) {
 		iface->radar_bit_pattern = radar_bitmap;
 		iface->conf->ru_punct_bitmap = cur_punct_bits;
-		u8 oper_centr_freq_seg0_idx = iface->conf->vht_oper_centr_freq_seg0_idx;
-		u8 oper_centr_freq_seg1_idx = iface->conf->vht_oper_centr_freq_seg1_idx;
+		u8 oper_centr_freq_seg0_idx = hostapd_get_oper_centr_freq_seg0_idx(iface->conf);
+		u8 oper_centr_freq_seg1_idx = hostapd_get_oper_centr_freq_seg1_idx(iface->conf);
 
 		chan_width = convert_to_oper_chan_width(chan_width);
 
--- a/src/ap/ieee802_11_vht.c
+++ b/src/ap/ieee802_11_vht.c
@@ -107,20 +107,13 @@ u8 * hostapd_eid_vht_operation(struct ho
 	os_memset(oper, 0, sizeof(*oper));
 
 #ifdef CONFIG_IEEE80211BE
-	if (hapd->iconf->punct_bitmap) {
-		punct_update_legacy_bw(hapd->iconf->punct_bitmap,
+	if (hapd->iconf->ru_punct_bitmap) {
+		punct_update_legacy_bw(hapd->iconf->ru_punct_bitmap,
 				       hapd->iconf->channel,
 				       &oper_chwidth, &seg0, &seg1);
 	}
 #endif /* CONFIG_IEEE80211BE */
 
-	if (hapd->iconf->ru_punct_bitmap)
-		punct_update_legacy_bw(hapd->iconf->ru_punct_bitmap,
-				   hapd->iconf->channel,
-				   &hapd->iconf->vht_oper_chwidth,
-				   &hapd->iconf->vht_oper_centr_freq_seg0_idx,
-				   &hapd->iconf->vht_oper_centr_freq_seg1_idx);
-
 	/*
 	 * center freq = 5 GHz + (5 * index)
 	 * So index 42 gives center freq 5.210 GHz
